File: hipstore/src/main/java/tech/lab23/hipstore/EntitiesStorage.java
package tech.lab23.hipstore;

import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class EntitiesStorage<T> implements Hipstore.MultiEntities<T> {

    private final Class<T> clazz;
    private final String key;
    private final Gson gson = new Gson();
    private final JsonParser jsonParser = new JsonParser();
    private final SharedPreferences prefs;

    public EntitiesStorage(SharedPreferences prefs, Class<T> clazz) {
        this.clazz = clazz;
        this.key = clazz.getSimpleName();
        this.prefs = prefs;
    }

    public EntitiesStorage(SharedPreferences prefs, Class<T> clazz, String storageKey) {
        this.clazz = clazz;
        this.key = storageKey;
        this.prefs = prefs;
    }

    @Override
    public boolean contains(T item) {
        return getAll().contains(item);
    }

    @Override
    public void remove(T item) {
        final List<T> list = getAll();
        list.remove(item);
        prefs.edit().putString(key, listToJsonString(list)).apply();
    }

    @Override
    public void add(T item) {
        final List<T> list = getAll();
        list.add(item);
        prefs.edit().putString(key, listToJsonString(list)).apply();
    }

    @Override
    public void add(T item, int index) {
        final List<T> list = getAll();
        if (index > list.size()-1) {
            add(item);
        } else {
            list.add(index, item);
        }
        prefs.edit().putString(key, listToJsonString(list)).apply();
    }

    @Override
    public List<T> getAll() {
        final String jsonString = prefs.getString(key, "");
        return jsonStringToList(jsonString);
    }

    @Override
    public void clear() {
        prefs.edit().remove(key).apply();
    }

    private String listToJsonString(List<T> list) {
        return gson.toJson(list);
    }

    private List<T> jsonStringToList(String jsonString) {
        final JsonElement jsonElement = jsonParser.parse(jsonString);
        if (jsonElement.isJsonArray()) {
            final JsonArray jsonArray = jsonElement.getAsJsonArray();
            final List<T> list = new ArrayList<>(jsonArray.size());
            for (JsonElement aJsonArray : jsonArray) {
                final T item = gson.fromJson(aJsonArray, clazz);
                list.add(item);
            }
            return list;
        }
        return new ArrayList<>(0);
    }
}

File: hipstore/src/main/java/tech/lab23/hipstore/EntitiesStorage.kt
package tech.lab23.hipstore

import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.JsonArray
import com.google.gson.JsonParser
import java.io.Serializable
import java.util.*

class EntitiesStorage<T : Serializable>(val prefs: SharedPreferences, val clazz: Class<T>) : Hipstore.MultiEntities<T> {
    private val KEY = clazz.simpleName
    private val gson = Gson()
    private val jsonParser = JsonParser()

    override fun contains(item: T): Boolean {
        return getAll().contains(item)
    }

    override fun remove(item: T) {
        val list = getAll()
        list.remove(item)
        prefs.edit().putString(KEY, listToJsonString(list)).apply()
    }

    override fun add(item: T) {
        val list = getAll()
        list.add(item)
        prefs.edit().putString(KEY, listToJsonString(list)).apply()
    }

    override fun getAll(): MutableList<T> {
        val jsonString = prefs.getString(KEY, "")
        return jsonStringToList(jsonString)
    }

    override fun clear() {
        prefs.edit().remove(KEY).apply()
    }

    private fun listToJsonString(list: List<T>) : String {
        return gson.toJson(list)
    }

    private fun jsonStringToList(jsonString: String): MutableList<T> {
        val jsonElement = jsonParser.parse(jsonString)
        if (jsonElement.isJsonArray) {
            val jsonArray: JsonArray = jsonElement.asJsonArray
            val list: MutableList<T> = ArrayList(jsonArray.size())
            val iterator = jsonArray.iterator()
            while (iterator.hasNext()) {
                val item: T = gson.fromJson(iterator.next(), clazz)
                list.add(item)
            }
            return list
        }
        return ArrayList<T>(0)
    }
}

